Support wider range of types in FutureNCCL

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* #48789 Make CUDAFuture remember and restore current device in callback
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* **#48502 Support wider range of types in FutureNCCL**
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

lw commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -430,20 +430,18 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
-    } else {
-      tensor = value.toTensor();
+    at::IValue::HashAliasedIValues sub_values;
+    // Prefer getSubValues() over visit() as the latter is a silent no-op for
+    // some unsupported types, whereas the former at least fails loudly.
+    value.getSubValues(sub_values);
```
To clarify this, it seems that getSubValues triggers an assert for these types:
https://github.com/pytorch/pytorch/blob/8b248af35d43c97d0e437f6f4ff0fbd4da5700c8/aten/src/ATen/core/ivalue.cpp#L202-L206
Whereas visit just ignores such values but succeeds silently. I must admit that I expected getSubValues and visit to behave consistently (and perhaps the former to be implemented through the latter). Is there any reason for that?

Also, are there fundamental reasons for those types to not be supported? Are they planned in the future?

Are there better ways to extract DataPtrs from an IValue? One idea I had was pickling it, but I'm worried about the performance implications.

mrshenli replied:
cc @wanchaol :)

mrshenli replied:
> Are there better ways to extract DataPtrs from an IValue? One idea I had was pickling it, but I'm worried about the performance implications.

I assume you are concerned about nested data structures with Tensors? It should be fine for `ProcessGroupNCCL` and futures returned by RPC, but might need a recursive/pickle-like solution for general `CudaFuture` objects returned by then?

lw replied:
Two cases that I'm not sure are supported in this version of the code are mixed-type lists (e.g., the user having a callback that does `return [42, tensor]`) and custom user-defined classes that contain tensors as some fields.

Also, I am not sure how this applies to RPC, because in that case the IValue held by the future will be a `Message`, which will contain the result of the remote user function as a field, and that value could be of any type, including a mixed-type lists and user-defined classes. Right?

lw replied:
As a general comment, if `getSubValues()` or `visit()` was able to support PyObjects (and properly acquire the GIL and such), then we wouldn't need all that DataPtr extractor craziness. That would by far be the ideal scenario. I don't know how much effort is required to get there...

wanchaol replied:
> admit that I expected getSubValues and visit to behave consistently (and perhaps the former to be implemented through the latter). Is there any reason for that?

semantic-wise it seems these two functions are a bit different, but yeah i agree they could be rewrite together. cc @bzinodev  who implemented those two functions.


> are there fundamental reasons for those types to not be supported? Are they planned in the future?

I think it's a matter of semantic undefined for those kind of types, i.e. what does it mean to get a subvalues of a `FutureType` ivalue? there might not be subvalues inside if the future is not completed. If we could defined well on those semantics, we could add those. 

> Are there better ways to extract DataPtrs from an IValue?

I think what you need here is a "flatten" function which flattens all different types of objects so that you could iterate them and get all the tensor data ptrs for those "tensor" ivalues. Your approach looks good by using `getSubValues`, though we still need to define what it should behave for those unsupported types if we want to expand it.

 > As a general comment, if getSubValues() or visit() was able to support PyObjects (and properly acquire the GIL and such), then we wouldn't need all that DataPtr extractor craziness.

Can you explain a bit more when/why we need to extract PyObjects to get DataPtrs? not sure if I understood fully on why supporting PyObjects would remove those craziness. But you want a python land to iterate the ivalues, you can possibly define those inside https://github.com/pytorch/pytorch/blob/2e9d6d99be52cdd2f27512ff357e8aae7f4d3d78/torch%2Fcsrc%2Fjit%2Fpython%2Fpython_ivalue.h#L0-L1

lw replied:
Thanks for your answers!

> I think it's a matter of semantic undefined for those kind of types

Good point, makes sense! I'd say I'm totally fine with only supporting a (reasonable) subset of types in CUDAFuture. The remaining concern is how can we effectively communicate this to the user? Both in terms of how do we raise a meaningful error, and do we explain it in the docs? Do you have suggestions?

> Can you explain a bit more when/why we need to extract PyObjects to get DataPtrs?

Sorry, it's the other way around: I need to extract DataPtrs from PyObjects. The reason is that the value of a "follow-up" future could be produced by a user-provided Python function, in which case the type of the IValue would be PyObject, right?

Anyways, thanks a lot for pointing me to python_ivalue.h! I hadn't seen it before and it gave me an idea to fix this properly. I'll try to implement and send out a PR, and maybe we can discuss further over there.

lw replied:
Done this in #48840.

lw commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -128,9 +130,10 @@ struct VISIBILITY_HIDDEN PythonFutureWrapper
-        // FIXME Should we support more types than just tensor lists?
-        auto new_value = torch::jit::toIValue(
-            obj, c10::ListType::create(c10::TensorType::get()));
+        // FIXME This could fail. As a fallback we could try to pickle the
+        // object, since the pickler might support broader types and it is able
+        // to extract the tensors from the object as a vector.
+        auto new_value = torch::jit::toTypeInferredIValue(obj);
```
My understanding is that the type can only be inferred when it's "regular", i.e., a mixed-type list won't work. Are there ways to handle such cases? Would pickling work?

More in general, I'd like to understand more about the PyObject type. Are all objects coming from Python represented as PyObjects, or only the ones for which no better type is found? Do we always need to hold the GIL when manipulating such types?

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]

wanchaol commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -430,20 +430,18 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
   // Expose the default implementation so that external ones can defer to it.
   static std::vector<std::reference_wrapper<const at::DataPtr>>
   defaultDataPtrExtractor(const at::IValue& value) {
```
seems to me that `defaultDataPtrExtractor` is only used by pybind_utils.h, maybe we can move this to `python_ivalue.h` to make it handle the PyObject case as well?

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good

lw committed:
Update on "Support wider range of types in FutureNCCL"


This commit is part of a stack that reworks FutureNCCL in order to extract a generic CUDA-aware Future subclass. The stack deliberately breaks up this transition into elementary changes, to make it easier to verify that the behavior is preserved (or to highlight how it gets changed).

---

FutureNCCL restricted the values to be tensors, or (singleton) lists of tensors, or Python object that could be converted to either of those types. We need a CUDA future that can handle more generic types though.

The main challenge is extracting all DataPtrs from an arbitrary object. I think I found some ways of doing so, but I'd like some JIT experts to look into this and tell me if there are better ways. I'll add inline comments for where their input would be appreciated.

Differential Revision: [D25177562](https://our.internmc.facebook.com/intern/diff/D25177562/)

[ghstack-poisoned]
